### DISCUSSION TITLE
FlagDAO —— Smart Contract on zksync

### DISCUSSION THREAD

**Discussion by Demian101** [zkp-co-learning/zkp-co-learn#99] [I've got an idea!]
ideas 提供者：大力 Lee，孝羽

### FlagDAO

FlagDAO 是一个为爱立 Flag 星人准备的 DAO，在本 DAO 中，每个人都有他自己的 Flag，每个人都想实现他的 Flag，每个人都会为其 Flag 进行质押，或者尝试判断别人的 Flag 什么时候倒 （：

我们的共识是：

1. 谨慎乐观投资自己
2. 嘻嘻哈哈围观他人

### How？

本次将为 FlagDAO 实现一个小的合约

- 目的：本次合约的目的是为【Rust 共学编程】实现质押合约
- 开发：使用 Solidity 部署到 **zksync**

学员 A 的 FLAG：学完这个自己开发一条小型公链
学员 A 会有一个 Notion 自我介绍的 页面，别人可以根据他的自我介绍，结合项目难度，去总体评估这个 Flag

### Roadmap

> $CT:  Co-learn Token

- 充值 399 RMB（链下） ---  根据白名单发放 10000 $CT Token（链上） (**1)
- 获得 $CT 后需要：
    - Create 自己的 Flag，输入文本并输入质押资金数量 $CT
    - 如果还有剩余的 $CT, 可以选择投资其他人的 Flag，输入其地址，输入质押资金数量
- 围观其他人的 Flag 时，可以根据其个人情况、Flag 情况综合评估其 Flag 会不会倒，进而做出投资判断；如果该 Flag 最终轰然崩塌，围观者可以收回自己的 $CC ，并根据质押比例均分其 $CC；
- 或者不将其当做投资，而是单纯以次来激励 flager，也可以进行质押，
- 如何判断 Flag 的完成？(**2)
    - 暂定由值得信赖和大家认可的多签账户来评估和更改合约状态

- 质押赎回后，可以等值参与到下一次共学（ 每 $10000 CC 最高可抵扣 ￥399 的 co-learn 课程（No，只是共学点返还，其他方式不可返还。））

> 每个人的 Flag 会有一个完成周期（如一个半月），到期后，如果 Flag 状态不为 Done，则均分给围观质押群众；如果没有人围观&质押，则自动销毁

(**1) 这笔 $CT 可以投资自己，全额投资自己的 flag，最高可以收回本金(.. )； 也可以投资别人的 flag，或许可以获得超额收益
(**2) 这部分的中心化评估是必要的，链下评估，链上改变其合约状态。

### Token

基础信用分 600，>= 600 才能发 Flag，不可转让。

三代币模型：
- NFT（SBT）
    - 投资人（立 Flag 成功 —— 信用评级）
    - 旗子 🚩 倒了
- 功能性代币 —— 信用分（项目评估的系数）
    - 积分比 SBT 更好用： 
        - 比如 Demian 信用积分是 800，小明只能 1:1 质押 
        - 孝羽的信用积分是 1000，小明只能 1.2:1 质押，孝羽能获取到更多的钱 💰 

- 证券型代币 —— 质押&流动性

为什么不质押 USDT？

- MVI（最有价值交互）
   - DAO 内悬赏通过 Flag 质押增加 DAO 内协作效率
   - 早期项目方通过 flag 质押解决无背景的筹资信任问题

### 招募：

1. 前端开发工程师一名
2. 合约开发工程师各一名

GAS： 给共学的组织者一些 Gas

验证其 Flag 完成， burn 一部分 Token > 国库
- 合格(50%)、优秀（90% 返还) 、Perfect(100% 返还)
- 共学即是运动员又是裁判员

- FlagDAO 变成公证处。质押的收益给第三方验证者发劳务费。

### 【TODO】

 - [ ] 多签评估 Flag 的完成，有没有更好更 easy 的方法

立 Flag 成功的 POAP、失败的 POAP

**Comment by Lipincode** [comment]
thks
